gateway-service/src/main/java/org/zowe/apiml/gateway/conformance/VerificationOnboardService.java

```diff
@@ -43,18 +53,82 @@ public boolean checkOnboarding(String serviceId) {
 
 
     /**
-     * Accepts metadata and retrieves the Swagger url if it existsd
+     * Accepts metadata and retrieves the Swagger url if it exists
      *
      * @param metadata to grab swagger from
      * @return SwaggerUrl when able, empty string otherwise
      */
-    public String retrieveSwagger(Map<String, String> metadata) {
-        String swaggerUrl = metadata.get("apiml.apiInfo.api-v2.swaggerUrl");
+    public String findSwaggerUrl(Map<String, String> metadata) {
+
+        String swaggerKey = null;
+        for (String key : metadata.keySet()) {
+            if (key.contains("swaggerUrl")) {        // Find the correct key for swagger docs, can be both apiml.apiInfo.0.swaggerUrl or apiml.apiInfo.api-v1.swaggerUrl for example
+                swaggerKey = key;
+                break;
+            }
+        }
+        if (swaggerKey == null) {
+            return "";
+        }
+        String swaggerUrl = metadata.get(swaggerKey);
         if (swaggerUrl != null) {
             return swaggerUrl;
         }
         return "";
     }
 
 
+    /**
+     * Retrieves swagger from the url
+     *
+     * @param swaggerUrl URL to retrieve from
+     * @return Swagger as string
+     */
+    public String getSwagger(String swaggerUrl) {
+        String response;
+        response = restTemplate.getForEntity(swaggerUrl, String.class).getBody();
+        return response;
+    }
+
+    /**
+     * Checks if endpoints can be called and return documented responses
+     *
+     * @param getEndpoints GET endpoints to check
+     * @return List of problems
+     */
+    public List<String> testGetEndpoints(Set<Endpoint> getEndpoints) {
+        ArrayList<String> result = new ArrayList<>();
+
+        for (Endpoint endpoint : getEndpoints) {
+            String urlFromSwagger = endpoint.getUrl();
+            // replaces parameters in {} in query
+            String url = urlFromSwagger.replaceAll("\\{[^{}]*}", "dummy");
+
+            ResponseEntity<String> response;
+            try {
+                response = restTemplate.getForEntity(url, String.class);
+            } catch (HttpClientErrorException | HttpServerErrorException e) {
+                response = ResponseEntity.status(e.getRawStatusCode())
+                    .headers(e.getResponseHeaders())
+                    .body(e.getResponseBodyAsString());
+            }
+
+            String responseBody = response.getBody();
+
+            if (responseBody != null && response.getStatusCode() == HttpStatus.NOT_FOUND && responseBody.contains("ZWEAM104E")) {
+                result.add("Documented endpoint at " + endpoint.getUrl() + " could not be located, attempting to call it through gateway gives the ZWEAM104E error");
+            }
+
+            if (!(endpoint.getValidResponses().get("GET").contains(String.valueOf(response.getStatusCode().value())) || endpoint.getValidResponses().get("GET").contains("default"))) {
```

Comment:
This if expression is quite complicated to read, I'd suggest to extract it to a method, and split it.
Also, what does it mean to have a "default" value in the valid responses?

Reply:
The getValidResponses() function returns a set of documented responses in the Swagger doc, "default" basically means that the endpoint can return any response. https://swagger.io/docs/specification/describing-responses/